transport/proxy.go

```diff
@@ -107,7 +108,11 @@ func doHTTPConnectHandshake(ctx context.Context, conn net.Conn, addr string) (_
 	}
 	resp.Body.Close()
 	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("failed to do connect handshake, status code: %s", resp.Status)
+		dump, err := httputil.DumpResponse(resp, true)
```

Comment:
Did you test this at all? Because there is the `resp.Body.Close()` line right before the if.